File: leetcode/Python/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        mn, mx = min(p.val, q.val), max(p.val, q.val)
        while root.val < mn or root.val > mx:
            root = root.left if root.val>mx else root.right
        return root

File: leetcode/C++/lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int mn=p->val,mx=q->val;
        if(p->val>q->val){mn=q->val; mx=p->val;}
        while(root->val < mn || root->val > mx){
            root = (root->val>mx)? root->left : root->right;
        }
        return root;
    }
};